### PR TITLE
Suport host rewrite

### LINKED ISSUE #1202: [Feature] Support host rewrite
### The New Feature

Describe the new feature you want to support clearly.

### Your scenes

Describe your use scenes (why need this feature).

### Your advice

Describe the advice or solution of this new feature.

### Environment

- MOSN Version

### PR DESCRIPTION
**Pull request by liangyuanpeng**
### Issues associated with this PR
closed #1202
Your PR should present related issues you want to solve.

### Solutions

当前是直接在base_rule.go#finalizeRequestHeaders中直接扩展  

原先:
```
...
if len(rri.hostRewrite) > 0 {
		headers.Set(protocol.IstioHeaderHostKey, rri.hostRewrite)
	} 
```

现在:
```
if len(rri.hostRewrite) > 0 {
		headers.Set(protocol.IstioHeaderHostKey, rri.hostRewrite)
	} else if rri.autoHostRewrite && configmanager.IsDnsTypeCluster(rri.ClusterName()) {
		headers.Set(protocol.IstioHeaderHostKey, rri.ClusterName())
	} else if len(rri.autoHostRewriteHeader) > 0 {
		if headerValue, ok := headers.Get(rri.autoHostRewriteHeader); ok {
			headers.Set(protocol.IstioHeaderHostKey, headerValue)
		}
	}
```
在configmanager中添加了`IsDnsTypeCluster` 
```
func IsDnsTypeCluster(clusterName string) bool {
	for _, cluster := range config.ClusterManager.Clusters {
		if cluster.ClusterType == v2.STRICT_DNS_CLUSTER || cluster.ClusterType == v2.LOGICAL_DNS_CLUSTER {
			return true
		}
	}
	return false
}
```
只有在Cluster类型为`STRICT_DNS`和`LOGICAL_DNS`的情况下才会执行,而判断是否为这两种类型是否符合条件是直接遍历当前所有`Clusters`,  

但是我认为维护一个map存储当前符合条件的cluster会比较好: 
`dnsClusterMap  key: clusterName value:1  `

这里实现为循环是因为维护一个这样的map涉及xds对clusters列表的修改, 在clusters列表在被修改时如果`STRICT_DNS`和`LOGICAL_DNS`的cluster被修改了,这个map同时也需要更新.  

而当前xds在另一个PR中处于WIP,并且我还对XDS涉及到的这部分内容还不是很熟悉,所以这里只做了个简单实现.(从性能上讲存在缺陷)

### UT result

```
Running tool: /usr/local/go/bin/go test -timeout 30s mosn.io/mosn/pkg/router -run ^(Test_RouteRuleImplBase_FinalizeRequestHeaders)$

ok  	mosn.io/mosn/pkg/router	(cached)
```

### Benchmark

If your code involves the processing of every request, you should give the Benchmark Result.

当前有压测工具吗?

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

golint结果显示当前库已有代码似乎挺多都不符合,列出少部分： 

```
../../../pkg/configmanager/configmanager.go:25:1: comment on exported function AddOrUpdateClusterConfig should be of the form "AddOrUpdateClusterConfig ..."
../../../pkg/configmanager/configmanager.go:60:1: exported function RemoveClusterConfig should have comment or be unexported
../../../pkg/configmanager/configmanager.go:86:1: exported function DeleteClusterHost should have comment or be unexported
../../../pkg/configmanager/configmanager.go:112:1: exported function AddOrUpdateClusterHost should have comment or be unexported
../../../pkg/configmanager/configmanager.go:135:1: comment on exported function AddPubInfo should be of the form "AddPubInfo ..."
```  

当前是如何处理这部分的？

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
=== RUN   TestListenerStart
2020-06-07 16:08:22,405 [INFO] [network] [listener start] [accept] listener test_listener 127.0.0.1:10101 closed
2020-06-07 16:08:24,406 [INFO] [network] [listener start] test_listener restart listener 
--- PASS: TestListenerStart (4.00s)
    listener_test.go:57: dial error: dial tcp 127.0.0.1:10101: connect: connection refused
    listener_test.go:57: dial error: dial tcp 127.0.0.1:10101: connect: connection refused
FAIL
coverage: 26.6% of statements
FAIL	mosn.io/mosn/pkg/network	25.038s
make: *** [Makefile:33: coverage-local] Error 1
make: *** [coverage] Error 2
The command "make coverage" exited with 2.

**Comment by liangyuanpeng** [comment]
@wangfakang  PTAL

File: pkg/router/base_rule.go [hunk 1]
```diff
@@ -202,6 +205,12 @@ func (rri *RouteRuleImplBase) finalizeRequestHeaders(headers api.HeaderMap, requ
 	rri.vHost.globalRouteConfig.requestHeadersParser.evaluateHeaders(headers, requestInfo)
 	if len(rri.hostRewrite) > 0 {
 		headers.Set(protocol.IstioHeaderHostKey, rri.hostRewrite)
+	} else if rri.autoHostRewrite && configmanager.IsDnsTypeCluster(rri.ClusterName()) {
```
Comment (wangfakang) [inline pkg/router/base_rule.go:208]: `autoHostRewrite` means that the host header will be swapped with the hostname of the upstream host chosen by the cluster manager and not `ClusterName`.

File: pkg/configmanager/configmanager.go [hunk 2]
```diff
@@ -414,3 +414,12 @@ func RmMqConsumers(key string) {
 
 	dump(true)
 }
+
+func IsDnsTypeCluster(clusterName string) bool {
+	for _, cluster := range config.ClusterManager.Clusters {
```
Comment (wangfakang) [inline pkg/configmanager/configmanager.go:419]: could refer `pkg/upstream/cluster/cluster_manager.go` file's   `cm.clustersMap.Load(clusterName)`.

File: pkg/router/base_rule.go [hunk 3]
```diff
@@ -18,17 +18,19 @@
...
 	"time"
 
 	"mosn.io/api"
-	"mosn.io/mosn/pkg/config/v2"
+	v2 "mosn.io/mosn/pkg/config/v2"
 	"mosn.io/mosn/pkg/log"
 	"mosn.io/mosn/pkg/protocol"
 	httpmosn "mosn.io/mosn/pkg/protocol/http"
 	"mosn.io/mosn/pkg/types"
+	"mosn.io/mosn/pkg/upstream/cluster"
```
Comment (wangfakang) [inline pkg/router/base_rule.go:33]: caused import cycle https://travis-ci.com/github/mosn/mosn/jobs/347205595

Comment (liangyuanpeng) [reply pkg/router/base_rule.go:33]: I have run command and nothing changed , what can i do for it ? 

Thank you!

```
goimports -w pkg/router/base_rule.go  
goimports -w pkg/router/base_rule_test.go
```

Comment (wangfakang) [reply pkg/router/base_rule.go:33]: > I have run command and nothing changed , what can i do for it ?
> 
> Thank you!
> 
> ```
> goimports -w pkg/router/base_rule.go  
> goimports -w pkg/router/base_rule_test.go
> ```

It is the package import cycle that causes the [CI task]( https://travis-ci.com/github/mosn/mosn/jobs/347205595) to fail, as  below

```
--------Run test package: mosn.io/mosn/pkg/upstream/cluster
# mosn.io/mosn/pkg/upstream/cluster
import cycle not allowed in test
package mosn.io/mosn/pkg/upstream/cluster (test)
	imports mosn.io/mosn/pkg/router
	imports mosn.io/mosn/pkg/upstream/cluster
FAIL	mosn.io/mosn/pkg/upstream/cluster [setup failed]
make: *** [Makefile:33: coverage-local] Error 1
make: *** [coverage] Error 2
The command "make coverage" exited with 2.
```

Comment (liangyuanpeng) [reply pkg/router/base_rule.go:33]: I tried to solve this `import cycle not allowed in test` , but it not working.   

I am not very skilled of golang , do you have any suggestions ? 

Thank you for your help!

Comment (wangfakang) [reply pkg/router/base_rule.go:33]: It has been resolved.

**Pull request merged**